File: C++/BinaryTree/preorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        if(root == NULL)
            return res;
        res.push_back(root->val);

        vector<int> l = preorderTraversal(root->left);
        vector<int> r = preorderTraversal(root->right);

        res.insert(res.end(), l.begin(), l.end());
        res.insert(res.end(), r.begin(), r.end());
        
        return res;
    }
};

File: Python3/BinaryTree/preorderTraversal.py
# -*- coding:UTF-8 -*-
'''
Solution：前序遍历
- 前序遍历顺序：根，左，右
- 利用一个栈（先进后出），把根节点先放进去，
    - 然后弹出根节点，再放根节点的右节点，然后是左节点；
    - 下次再弹出左节点，然后再放它的右节点，左节点； 
'''
# # Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        
        res, stack = [], []
        stack.append(root)
        while(stack):
            temp = stack.pop()
            res.append(temp.val)
            if temp.right:
                stack.append(temp.right)
            if temp.left:
                stack.append(temp.left)
        return res
        
        
        
        